cpp/leaky_core.cc

```diff
@@ -252,8 +245,8 @@ Str* BackslashEscape(Str* s, Str* meta_chars) {
     }
     *p++ = c;
   }
-  int len = p - buf;
-  return StrFromC(buf, len);
+  int len = p - buf->data_;
+  return StrFromC(buf->data_, len);
```

Comment:
There's an API called OverAllocatedStr() that encapsulates the pattern of:

- allocating an upper bound
  -  s = OverAllocatedStr(n)
- shrinking to fit after you know how many bytes were written
  -  s->SetObjLenFromStrLen(10)

I used it for read(), strftime() and a similar style for gethostname(), getcwd()   (grep should easily reveal them)

This will avoid copying the buffer -- e.g. StrFromC() copies and takes ownership

----

Honest question (looking for feedback): what do you think of this API?   Jesse found it to be error prone, while I think it is a natural bridge between C-style buffer interfaces and C++ GC objects, which avoids some copying

Reply:
Ah great! I was a little bummed by the double alloc here. Glad there's a way to avoid.

I think the API is reasonable. Not really sure how else you'd do something like this without allocating twice, and I assume the wasted bytes from the shrink (i assume it's not actually giving back the slop) will be minimal and short lived. The name for the shrink op is a little awkward though. What if it were something like `s->Truncate(...)`, `s->Trim(...)`, or `s->Adjust(...)`?